File: src/main/java/org/taktik/icure/logic/impl/filter/patient/PatientByHcPartyNameFilter.java
package org.taktik.icure.logic.impl.filter.patient;

import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.entities.Patient;
import org.taktik.icure.entities.User;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.logic.PatientLogic;
import org.taktik.icure.logic.impl.filter.Filter;
import org.taktik.icure.logic.impl.filter.Filters;

import javax.security.auth.login.LoginException;
import java.util.HashSet;
import java.util.Set;

public class PatientByHcPartyNameFilter implements Filter<String, Patient, org.taktik.icure.dto.filter.patient.PatientByHcPartyNameFilter> {

    PatientLogic patientLogic;
    ICureSessionLogic sessionLogic;

    @Autowired
    public void setPatientLogic(PatientLogic patientLogic) {
        this.patientLogic = patientLogic;
    }
    @Autowired
    public void setSessionLogic(ICureSessionLogic sessionLogic) {
        this.sessionLogic = sessionLogic;
    }

    private String getLoggedHealthCarePartyId() throws LoginException {
        User user = sessionLogic.getCurrentSessionContext().getUser();
        if (user == null || user.getHealthcarePartyId() == null) {
            throw new LoginException("You must be logged to perform this action. ");
        }
        return user.getHealthcarePartyId();
    }

    @Override
    public Set<String> resolve(org.taktik.icure.dto.filter.patient.PatientByHcPartyNameFilter filter, Filters context) {
        try {
            return new HashSet<>(patientLogic.listByHcPartyName(
                filter.getName(),
                filter.getHealthcarePartyId() != null ? filter.getHealthcarePartyId() : getLoggedHealthCarePartyId()));
        } catch (LoginException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/filter/patient/PatientByHcPartyNameFilter.kt
package org.taktik.icure.asynclogic.impl.filter.patient

import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.PatientLogic
import org.taktik.icure.asynclogic.impl.filter.Filter
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.domain.filter.patient.PatientByHcPartyNameFilter
import org.taktik.icure.entities.Patient
import org.taktik.icure.utils.getLoggedHealthCarePartyId
import javax.security.auth.login.LoginException

class PatientByHcPartyNameFilter(private val patientLogic: PatientLogic,
                                 private val sessionLogic: AsyncSessionLogic) : Filter<String, Patient, PatientByHcPartyNameFilter> {

    override fun resolve(filter: PatientByHcPartyNameFilter, context: Filters) = flow<String> {
        try {
            emitAll(patientLogic.listByHcPartyName(filter.name, filter.healthcarePartyId ?: getLoggedHealthCarePartyId(sessionLogic)))
        } catch (e: LoginException) {
            throw IllegalArgumentException(e)
        }
    }
}
